### PR TITLE
[ext_authz]: ext_authz filter unit test that use real threading

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
Commit Message: Add test with real threading as a  followup of #17619 
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by chaoqin-li1123** [comment]
Have been able to reproduce the race condition for 
SUMMARY: ThreadSanitizer: data race /proc/self/cwd/test/mocks/grpc/mocks.cc:14:18 in Envoy::Grpc::MockAsyncClient::MockAsyncClient()::$_0::operator()(absl::string_view, absl::string_view, std::__1::unique_ptr<Envoy::Buffer::Instance, std::__1::default_delete<Envoy::Buffer::Instance> >&&, Envoy::Grpc::RawAsyncRequestCallbacks&, Envoy::Tracing::Span&, Envoy::Http::AsyncClient::RequestOptions const&) const

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:49]: can you merge these 3 loops into 1?

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:76]: I think the BlockikngCounter may be a good re-use here. See the example in thread_local_store_test.cc

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:83]: ABSL_GUARDED_BY(mutex_)

...though that won't be needed if you use a blocking counter.

File: test/mocks/grpc/mocks.h
Comment (jmarantz) [inline test/mocks/grpc/mocks.h:93]: I know due to our IM that you are only putting send_cnt here so that TSAN can detect a race on it if it's not properly threaded.

But you need to comment this strategy in the code, so a future reader of this code doesn't see this as a write-only variable and just remove it.

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:121]: can some of the helper code in this test-class be factored out and shared with other test-classes in this file?

**Review by jmarantz — COMMENTED** [review]
I need to look more deeply at your threading semantics to understand how you are triggering a potential race, as it diverges a bit from the examples I pointed you to.

But the higher level question is: were you able to verify that this detects a race by reverting the PR that was merged earlier?

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (chaoqin-li1123) [reply test/extensions/filters/http/ext_authz/config_test.cc]: Update a version that use BlockingCounter

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:57]: this is way easier to read! Thanks!

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:63]: I think you might be able to combine `startThreading()` with` postWorkToAllWorkers()`. In both call-sites you are initializing a factory between those two, but it doesn't look like there'd be anything stopping you from initializing the factory before the call to `startThreading()`.

File: test/mocks/grpc/mocks.h
Comment (jmarantz) [reply test/mocks/grpc/mocks.h:96]: Also you should probably test send_cnt_ in the tests against an expected value, at the end.

**Review by jmarantz — COMMENTED** [review]
/wait

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (chaoqin-li1123) [reply test/extensions/filters/http/ext_authz/config_test.cc:63]: Makes sense to create the factory before threading. but the test class will be more flexible if we don't combine them. For example, we can post the same job several times.

Comment (jmarantz) [reply test/extensions/filters/http/ext_authz/config_test.cc:63]: Cool cool no doubt no doubt :)

But you weren't actually doing that :)

OK leave them separate then.

**Comment by jmarantz** [comment]
Anyway please confirm that you can get a TSAN error with this PR by reverting the PR that fixed the race, and let me know when all else is resolved.

/wait

**Comment by chaoqin-li1123** [comment]
> Anyway please confirm that you can get a TSAN error with this PR by reverting the PR that fixed the race, and let me know when all else is resolved.
> 
> /wait
yes, this is part of the race condition log.
WARNING: ThreadSanitizer: data race (pid=15)
  Write of size 4 at 0x7b2c00001548 by thread T6:
    #0 Envoy::Grpc::MockAsyncClient::MockAsyncClient()::$_0::operator()(absl::string_view, absl::string_view, std::__1::unique_ptr<Envoy::Buffer::Instance, std::__1::default_delete<Envoy::Buffer::Instance> >&&, Envoy::Grpc::RawAsyncRequestCallbacks&, Envoy::Tracing::Span&, Envoy::Http::AsyncClient::RequestOptions const&) const /proc/self/cwd/test/mocks/grpc/mocks.cc:14:18 (config_test+0x3f53467)

**Comment by chaoqin-li1123** [comment]
It seems difficult expect the send_cnt_ at the end of the test, because RawAsyncClient is a private member deeply nested inside the ex_authz filter, and can't be accessed directly.

**Comment by jmarantz** [comment]
Won't you get that same client object from the cache when you ask for it?

Of course you'd have to dynamic_cast it to the RawAsyncClient impl but that maybe seems ok for a test?

**Comment by chaoqin-li1123** [comment]
> Won't you get that same client object from the cache when you ask for it?
> 
> Of course you'd have to dynamic_cast it to the RawAsyncClient impl but that maybe seems ok for a test?
The mock cluster manager and mock async client manager in this test don't have thread local cache, we return a new object every time getOrCreateRawAsyncClient is called.

**Comment by jmarantz** [comment]
Sorry I misunderstood what you had done in this test.

Can we use real objects for the threading tests rather than mocks? If we aren't using real TLS at least then I don't think this is really testing what was intended.

**Comment by chaoqin-li1123** [comment]
We can't have a real cluster manager here, that's much too heavy. but using thread local cache here is possible, I will give it a try.

**Comment by jmarantz** [comment]
That might be OK. But tell me why we can't use a real cluster manager?

**Comment by chaoqin-li1123** [comment]
> That might be OK. But tell me why we can't use a real cluster manager?

Because if we don't use a mock async client, we will need to send and receive real grpc request and response in each thread, which will make this test a multi-thread integration test. I prefer to add thread local to the MockAsyncClientManager.

**Comment by jmarantz** [comment]
It's not obvious to me why sending/receiving real requests within a process is a problem.

I am wanting a multi-thread test. I'm not sure what it is that makes it an 'integration' test (or why that's necessarily bad).

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (yanavlasov) [inline test/extensions/filters/http/ext_authz/config_test.cc:143]: I think you will see TSAN errors in this expectation, since gmock's internal bookkeeping structures for tracking expectations are not thread safe.
I think you will need to actually create the async_client_manager that is not mocked.

**Comment by chaoqin-li1123** [comment]
Window ci failure finally fixed!!! We should post a callback to reset the async client manager before joining the thread because the destructor of async client manager will try to post a callback to clear the thread local slot..

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (chaoqin-li1123) [reply test/extensions/filters/http/ext_authz/config_test.cc:167]: We wouldn't have a race here if ON_CALL don't modify the state of the object.

**Comment by jmarantz** [comment]
Let me know when this is ready for another look.

/wait

**Comment by chaoqin-li1123** [comment]
it is ready, the api_compat test  failure is unrelated.

**Review by jmarantz — COMMENTED** [review]
Nice!

can we give a shot at doing the refactor here to share the multi-threaded-test mechanism with thread_local_store_test.cc ?

If another one of these needs to be added we don't want to go through this again!

Also there are a few other still unresolved comments.

/wait

File: test/mocks/grpc/mocks.h
Comment (chaoqin-li1123) [reply test/mocks/grpc/mocks.h:93]: Thanks, the send_cnt_ is now tested.

**Comment by chaoqin-li1123** [comment]
Thanks, will try to refactor.

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (adisuissa) [inline test/extensions/filters/http/ext_authz/config_test.cc:64]: `absl::Notification` probably makes more sense in this case.

File: test/mocks/grpc/mocks.h
Comment (adisuissa) [reply test/mocks/grpc/mocks.h:93]: +1 for a comment here
(was about to ask why it's needed)

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (adisuissa) [inline test/extensions/filters/http/ext_authz/config_test.cc:209]: nit: const

**Review by adisuissa — COMMENTED** [review]
Thanks for working on these tests!
Left a few comments.

**Comment by chaoqin-li1123** [comment]
I finish the refactoring, if both tests share the same header that includes implementation for RealThreadsTestBase, where should the header be?

**Comment by jmarantz** [comment]
I would put it in test/common.

**Comment by jmarantz** [comment]
Sorry that's wrong. I meant `test/test_common` :). test/common doesn't have any other code in it; just subdirs.

**Review by jmarantz — COMMENTED** [review]
Ping when ready for next round: this is getting a lot closer.

Can you go through all the comments also so we can close this out? Thanks!

/wait

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (chaoqin-li1123) [reply test/extensions/filters/http/ext_authz/config_test.cc]: Fixed

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (chaoqin-li1123) [reply test/extensions/filters/http/ext_authz/config_test.cc]: Thanks, removed.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1520]: nit: RealThreadsTestBase doesn't inherit from testing::Test so I wouldn't call it a base -- it's really a helper.

In fact I'm not sure you need to inherit from it: maybe you can just instantiate it as a member?

File: test/test_common/real_threads_test.h
Comment (jmarantz) [inline test/test_common/real_threads_test.h:9]: per above, s/RealThreadsTestBase/RealThreadsTestHelper/ ?

Also make the filename match the class name.

File: test/test_common/real_threads_test.h
Comment (jmarantz) [inline test/test_common/real_threads_test.h:23]: move body to cc file

File: test/test_common/real_threads_test.cc
Comment (jmarantz) [inline test/test_common/real_threads_test.cc:34]: It's more complex than desired, I see, because there are some slight differences in shutdown sequence between the two tests that use this.

Maybe you can leave this as is for now but leave a TODO to refactor slightly to avoid repeated code. E.g. maybe you could call a virtual hook to do the extra shutdown call needed in thread_local_store_test.cc but otherwise share the shutdownThreading impl which is almost identical.

I'd be fine having that as a TODO . The right structure is in place for sharing at least.

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_authz/config_test.cc:16]: are these 3 includes needed now? I think you get everything via RealThreadsTestBase.

File: test/test_common/real_threads_test.h
Comment (jmarantz) [inline test/test_common/real_threads_test.h:58]: make member vars private and access via API from the tests that use this helper.

**Review by jmarantz — COMMENTED** [review]
This is looking great! Just a few nits and I think this is ready to go, even though some more refactoring may still be desirable as a follow-up.

File: test/test_common/real_threads_test.h
Comment (chaoqin-li1123) [reply test/test_common/real_threads_test.h]: Thanks, fixed.

File: test/extensions/filters/http/ext_authz/config_test.cc
Comment (chaoqin-li1123) [reply test/extensions/filters/http/ext_authz/config_test.cc]: Thanks, removed.

File: test/test_common/real_threads_test.h
Comment (chaoqin-li1123) [reply test/test_common/real_threads_test.h]: It is tricky here because these unique_ptr members are managed(freed) by subclass because of the reason you mentioned above. So we can't just return a reference to the object here, but it also doesn't make much sense to me to return a reference to ptr in a getter.

File: test/test_common/real_threads_test.h
Comment (chaoqin-li1123) [reply test/test_common/real_threads_test.h]: Thanks, fixed.

File: test/common/stats/thread_local_store_test.cc
Comment (chaoqin-li1123) [reply test/common/stats/thread_local_store_test.cc]: Renamed. Technically making the helper class a member is possible, a base class is equivalent to a member initialized before all other members. But I think making it a base class sounds more intuitive to me because threading feels like an environment in which a test execute on instead of a member.

File: test/test_common/real_threads_test.cc
Comment (chaoqin-li1123) [reply test/test_common/real_threads_test.cc]: I was also thinking about a virtual override to customize the resource management. But this is intended to be called in dtor, virtual functions shouldn't be called in a destructor.

Comment (chaoqin-li1123) [reply test/test_common/real_threads_test.cc]: clang-tidy actually enforce that we don't call virtual function from dtor, my previous code result in CI failure.

Comment (jmarantz) [reply test/test_common/real_threads_test.cc]: Like I said I'm OK with a TODO to clean this up. This is clearly heading in the right direction.

As indicated above I think this might be better as a helper class rather than an inheritance.

Comment (chaoqin-li1123) [reply test/test_common/real_threads_test.cc]: Because we access many protected data members of the class directly, if we don't inherit from it, we need to make these members public or friend the class in each class that use it, which doesn't seem right to me.

Comment (jmarantz) [reply test/test_common/real_threads_test.cc]: Yeah this is an indication that the refactoring isn't shaped right yet. Let's just drop in a TODO and follow-up to clean it up.

**Review by jmarantz — COMMENTED** [review]
LGTM modulo adding a TODO to clean up the shared class and its interplay with the two uses.

/assign-from @envoyproxy/senior-maintainers

File: test/test_common/real_threads_test_helper.h
Comment (jmarantz) [inline test/test_common/real_threads_test_helper.h:41]: add doc for these functions.

File: test/test_common/real_threads_test_helper.h
Comment (jmarantz) [inline test/test_common/real_threads_test_helper.h:57]: add TODO to privatize these member vars.

**Review by jmarantz — COMMENTED** [review]
Looks great ; still some minor unaddressed comments.

Can you also re-check that if you revert the earlier threading fix, that this test catches the race?

**Review by jmarantz — COMMENTED** [review]
/wait

**Comment by chaoqin-li1123** [comment]
Comments addressed. Race condition is still detected after changes reverted.

**Review by jmarantz — APPROVED** [review]

File: test/mocks/grpc/mocks.h
Comment (snowp) [inline test/mocks/grpc/mocks.h:94]: nit: I'd make this `sent_count_` per the style guide (https://google.github.io/styleguide/cppguide.html#General_Naming_Rules)

File: test/test_common/real_threads_test_helper.h
Comment (snowp) [inline test/test_common/real_threads_test_helper.h:36]: `explicit`

**Review by snowp — CHANGES_REQUESTED** [review]
Some nits, otherwise LG

/wait

**Comment by jmarantz** [comment]
@chaoqin-li1123 something weird happened in coverage CI on the retest. I think maybe it's worth merging main to re-kick things.

**Comment by chaoqin-li1123** [comment]
Thanks, has merged main to retest.

**Review by jmarantz — APPROVED** [review]

**Pull request merged**
